news/2025-03-27-spectec/index.md

```diff
@@ -0,0 +1,205 @@
+---
+title: “SpecTec Has Been Adopted"
+date: 2025-03-27
+layout: news
+---
+# SpecTec Has Been Adopted
+
+_Published on March 27, 2025 by [Andreas Rossberg](https://github.com/rossberg)._
+
+Two weeks ago, the Wasm Community Group voted to adopt [“SpecTec”](https://github.com/Wasm-DSL/spectec/tree/main/spectec) for authoring future editions of the Wasm spec.
+In this post, I’ll shed some light on what SpecTec is, what it helps with, and why it takes Wasm to a new level of rigor and assurance that is unprecedented when it comes to language standards.
+
+
+### Formal Language Semantics
+
+One feature that sets Wasm apart from other mainstream programming technologies
+is that it comes with a complete formalisation:
+its syntax (binary & text format), type system (validation), and operational semantics (execution) are given in self-contained, mathematically precise terms.
+This formal specification was not an afterthought,
+it is an integral part of Wasm's design process and has been a normative part of the official language standard from day 1.
+And not just that: it enabled the *soundness* of the language
+— i.e., the fact that it has no undefined behaviour and no runtime type errors can occur —
+to be [machine-verified](https://github.com/WasmCert) before the first version of the standard was published.
+
+This was a huge leap forward, because the practical state of language specifications is basically stuck in the 1960s:
+most language standards, even new ones,
+are still defined by some basic grammar notation for their syntax (and sometimes not even that),
+while their semantics is given by a combination of pretty prose, hidden assumptions, and wishful thinking.
+
+It is hard to avoid ambiguities, contradictions, oversights, and serious bugs in such an approach.
+Given that programming languages are the very foundation
+upon which all our critical systems are built today,
+that is a bit scary.
+And not quite up to the standard that we came to expect from other mature engineering disciplines:
+imagine a suspension bridge built with the methodology of a tree house,
+from just pencil drawings and with statics guesstimated as we go.
+
+Fortunately, that gap is more or less a solved problem, at least in principle.
+When it comes to scientific papers and text books,
+programming languages are routinely defined by a combination of formal techniques developed in the 70s to 90s.
+In 1990 already, [The Definition of Standard ML](https://mitpress.mit.edu/9780262631327/the-definition-of-standard-ml/) was published,
+the first comprehensive formal definition of a general-purpose programming language of realistic size and scope. (In case you never have heard of it, ML was a precursor to contemporary functional languages like OCaml and Haskell, a highly influential academic language that pioneered features like type inference, type polymorphism (generics), variant types, pattern matching, and advanced module systems in the early 80s.)
```

Comment:
```suggestion
the first comprehensive formal definition of a general-purpose programming language of realistic size and scope. (In case you have never heard of it, ML was a precursor to contemporary functional languages like OCaml and Haskell, a highly influential academic language that pioneered features like type inference, type polymorphism (generics), variant types, pattern matching, and advanced module systems in the early 80s.)
```